File: machine learning models/SVM based Insurance Purchase Prediction/MATLAB/insurancesvm.m
function main()
	%% ================== Reset the workspace ===================
	clear all;close all;clc;

	%% ================== Insurance Dataset =====================
	[data,txt,raw] = xlsread('insurance.xls','data');
	y = double(strcmp(txt(2:end,31),'Yes')); % 0 and 1
	y(y==0)=-1; % 1 and -1
	X = [data(:,1:7),data(:,9:end)]; % only numeric fields
	X = normalize(X); % binary fields are not affected from min-max normalization
	[m,n] = size(X);
	beta = 0.95;

	% Stratified sampling, 2 strata
	train_pc = 0.7;
	train_ind = [randsample(1:sum(y==-1),floor(train_pc*sum(y==-1))),sum(y==-1)+randsample(1:sum(y==1),floor(train_pc*sum(y==1)))];
	TrainSize = size(train_ind,2);
	yTrain = y(ismember(1:m,train_ind),:);
	yTest = y(~ismember(1:m,train_ind),:);
	XTrain = X(ismember(1:m,train_ind),:);
	XTest = X(~ismember(1:m,train_ind),:);

	% New representation of the training and test data sets
	[XTrain,w] = pca(XTrain,beta);
	XTest=XTest*w; 

	% Gaussian Kernel
	sigma = 0.1;
	K = kernel(XTrain,XTrain,sigma,0);
	a0=eps*ones(TrainSize,1);

	% Inequality that individual alpha>=0
	A = -eye(TrainSize);
	b = zeros(TrainSize,1);

	% Equality that sum(alpha_i*y_i)=0
	Aeq = yTrain';
	beq = 0;

	C = 1e+20; 

	% Change from min to max optimization by multiplying with -1
	H = diag(yTrain)*K*diag(yTrain)+1e-10*eye(TrainSize); % Regularization term to force H positive definite
	f = -ones(TrainSize,1);

	options = optimset('UseParallel','always','Display','iter','MaxIter',100,'MaxFunEvals',1000000,'LargeScale', 'off');
	alpha = quadprog(H,f,A,b,Aeq,beq,zeros(TrainSize,1),C*ones(TrainSize,1),a0,options);

	indx = find(alpha >= eps);
	b = mean(yTrain(indx)-K(indx,:)*(alpha.*yTrain));
	pred = sign(((1+kernel(XTest,XTrain,sigma,0)).^2)*(alpha.*yTrain)+b);
	fprintf('Test Accuracy: %f\n', mean(double(pred==yTest))*100);

	pred = sign(((1+kernel(XTrain,XTrain,sigma,0)).^2)*(alpha.*yTrain)+b);
	fprintf('Train Accuracy: %f\n', mean(double(pred==yTrain))*100);

	function [J,grad] = costFunction(param,Y,r,n_lenders,n_loans,n_features,lambda)
		% Extract X and Theta from param vector
		X = reshape(param(1:n_loans*n_features),n_loans,n_features);
		Theta = reshape(param(n_loans*n_features+1:end),n_lenders,n_features);
					
		% Cost
		predictions = X*Theta'; % prediction,nm x nu
		errors = (predictions-Y).*r; % also nm x nu
		J = (1/2)*sum(sum(errors.^2));

		% Gradients
		X_grad = errors*Theta; % error is  nm x nu,and Theta is nu x n,X_grad is nm x n
		Theta_grad = errors'*X; % error' is  nu x nm,X is nm x n,so Theta_grad is nu x n

		% Regularized cost function to penalize overfitting
		reg_X = (lambda/2)*sum(sum(X.^2));
		reg_Theta = (lambda/2)*sum(sum(Theta.^2));
		J = J+reg_Theta+reg_X;

		% Add regularization terms to gradients
		X_grad = X_grad+lambda*X;
		Theta_grad = Theta_grad+lambda*Theta;

		grad = [X_grad(:); Theta_grad(:)];
	end
	
	function K=kernel(XTest,XTrain,sigma,type);
		% Gaussian kernel
		if type==0 
			K = XTest*XTrain';
		else 
			X1 = sum(XTest.^2,2);
			X2 = sum(XTrain.^2,2)';
			K = bsxfun(@plus,X1,bsxfun(@plus,X2,-2*XTest*XTrain'));
			K = exp(-K/2/sigma);
		end 
	end
	function [Xnorm] = normalize(X)
		[m,n] = size(X);
		maxval = max(X);
		minval = min(X);
		Xnorm = (X-repmat(minval,m,1))./(repmat(maxval,m,1)-repmat(minval,m,1));
	end
	function [pca,w,k]=pca(X,alpha)
		[n,d]=size(X);
		covm=cov(X);
		[E,D]=eig(covm); % can also use pca functions in Matlab for principal components
		[val,loc]=sort(diag(D),'descend');
		if alpha>=1
			k=alpha;
		else
			k=sum((cumsum(val)/sum(val))<=alpha);
			if k==0 k=1; end;
		end
		E=E(:,loc);
		w=E(:,1:k);
		length=sqrt(sum(w.^2));
		w=w./(ones(d,1)*length);
		pca=X*w; 

	end
end

File: machine learning models/SVM based Insurance Purchase Prediction/Python/insurancesvm.py
import svm_insurance_lib as lib
import csv
import numpy as np
import math
import cvxopt as opt
from cvxopt import matrix, solvers

# function that normalizes each predictor
def normalize(X):
    m = np.size(X, 0)
    # n = np.size(X, 1)
    maxval = np.max(X, 0)
    minval = np.min(X, 0)
    minMat = np.tile(minval, (m, 1))
    DifMat = np.tile(maxval, (m, 1)) - minMat
    Xnorm = np.divide(X - minMat, DifMat)
    return Xnorm

def pca(X, alpha):
    n = np.size(X, 1)
    covm = np.cov(X, rowvar = False)
    eig_val, eig_vec = np.linalg.eig(covm)   
    val = np.sort(eig_val)[::-1]
    loc = val.argsort()[::-1]
    if alpha >= 1:
        k = alpha
    else:
        k = np.sum(np.array(np.cumsum(val)/np.sum(val) <= alpha).astype(int))
        if k == 0:
            k=1
    E = eig_vec[:, loc]
    w = E[:, 0:k]
    length = np.sqrt(np.sum(w**2, axis=0))
    w = np.divide(w, np.dot(np.ones(n)[:, None], length.T[None,:]))
    return w
    
def kernel(XTest, XTrain, sigma, type):
    if type == 0:
        K = np.dot(XTest, XTrain.T)
    else:
        X1 = np.sum(XTest**2, axis = 1)
        X2 = np.sum(XTrain**2, axis = 1)
        foo1 = np.dot(np.ones(np.size(XTest, axis=0))[:, None], X2.T[None,:]) \
                      - 2 * np.dot(XTest, XTrain.T)
        K = np.dot(X1[:,None], np.ones(np.size(XTrain, axis=0)).T[None,:]) \
                   + foo1
        K = np.exp(-K/(2*sigma))
    return K   
        
# read data from csv file
datPath = r"C:\xxx\insurance.csv"
with open(datPath,'rb') as csvfile:
    csvData = csv.reader(csvfile)
    datList = []
    for row in csvData:
        datList.append(row)
        
# get the colnames in the first row and remove it
colNames = datList.pop(0)

# convert list to matrix
data = np.array(datList)

# insurance data set
y = np.array(data[:, 30], "str")

# index of X selected
idx = np.array([i for j in (range(7, 14), range(15, 29)) for i in j])
X = np.array(data[:, idx], "float")

# normalize X
X = lib.normalize(X)
m = np.size(X, 0)
n = np.size(X, 1)

# y is a 1d array
y = np.array([{"Yes": 1, "No": -1}[y[i]] for i in range(len(y))])

# Stratified sampling
train_pc = 0.7
size1 = math.floor(train_pc*sum(y == -1))
size2 = math.floor(train_pc*sum(y == 1))
idxY1 = np.linspace(0, len(y) - 1, len(y), dtype="int")[y == -1]
idxY2 = np.linspace(0, len(y) - 1, len(y), dtype="int")[y == 1]
B1train = np.concatenate((np.random.choice(idxY1, size1, False),
                          (np.random.choice(idxY2, size2, False))), axis=0)
yTrain = y[B1train]
yTest = np.delete(y, B1train, axis=0)
XTrain = X[B1train, :]
XTest = np.delete(X, B1train, axis=0)
TrainSize = np.size(XTrain, axis = 0)

# New representation of the training and test data sets
# reduce the X dimensions 
beta = 0.95
W = lib.pca(XTrain, beta)
XTrain = np.dot(XTrain, W)
XTest = np.dot(XTest, W)

# Gaussian Kernel
sigma = 0.1
K = lib.kernel(XTrain, XTrain, sigma, 0)
a0 = 2.0 * np.spacing(1) * np.ones(TrainSize)

# Inequality that individual alpha>=0
C = 1e+20
G = matrix(np.concatenate((-np.eye(TrainSize), np.eye(TrainSize)), axis = 0))
h = matrix(np.concatenate((np.zeros(TrainSize), C * np.ones(TrainSize)), axis = 0))

# Equality that sum(alpha_i*y_i)=0
A = opt.matrix(np.double(yTrain))
b = matrix(0.0)

# Change from min to max optimization by multiplying with -1
# Regularization term to force H positive definite
yTmat = np.diag(np.double(yTrain))
P = 0.5 * np.dot(yTmat, np.dot(K, yTmat)) + 1e-10 * np.identity(TrainSize)
q = opt.matrix(-np.ones((TrainSize,1)))

opts = {'abstol':1e-5, 'reltol':1e-5, 'maxiters':45}
sol = solvers.qp(matrix(P), q, G, h, A.T, b, initvals = a0, options=opts)
alpha = np.array(sol['x'])

b = np.mean(yTrain - np.dot(K, (np.multiply(alpha, yTrain))))
K1 = lib.kernel(XTest, XTrain, sigma, 0)
pred = np.sign(np.dot((1 + K1)**2, np.multiply(alpha.T, yTrain).T) + b)
print('Test accuracy: %f\n' % (np.mean(pred == yTest)*100))

K2 = lib.kernel(XTrain, XTrain, sigma, 0)
pred = np.sign(np.dot((1 + K2)**2, np.multiply(alpha.T, yTrain).T) + b)
print('Training accuracy: %f\n' % (np.mean(pred == yTrain)*100))
